lib/matplotlib/tests/test_colorbar.py

```diff
@@ -622,3 +622,84 @@ def test_colorbar_int(clim):
     im = ax.imshow([[*map(np.int16, clim)]])
     fig.colorbar(im)
     assert (im.norm.vmin, im.norm.vmax) == clim
+
+
+def test_anchored_cbar_position_using_specgrid():
+    data = np.arange(1200).reshape(30, 40)
+    levels = [0, 200, 400, 600, 800, 1000, 1200]
+    shrink = 0.5
+    anchor_y = 0.3
+    # right
+    fig, ax = plt.subplots()
+    cs = ax.contourf(data, levels=levels)
+    cbar = plt.colorbar(
+            cs, ax=ax, use_gridspec=True,
+            location='right', anchor=(1, anchor_y), shrink=shrink)
+
+    # the bottom left corner of one ax is (x0, y0)
+    # the top right corner of one ax is (x1, y1)
+    # p0: the verticle / horizontal postion of anchor
+    x0, y0, x1, y1 = ax.get_position().extents
+    cx0, cy0, cx1, cy1 = cbar.ax.get_position().extents
+    p0 = (y1 - y0) * anchor_y + y0
+
+    np.testing.assert_allclose(
+            [cy1, cy0],
+            [y1 * shrink + (1 - shrink) * p0, p0 * (1 - shrink) + y0 * shrink])
```

Comment:
I'm not sure I understand why the anchor point `p0` should be affected by `shrink`.

Reply:
`p0` is not affected by `shrink`, `cy1` is the top of colorbar axes, which can be calculated using `p0` and `shrink` 
`cy1 = y1 * shrink + (1 - shrink) * p0` 
while `p0` and `shrink` do not affect each other

Reply:
I'm not saying they affect _each other_; I'm saying that `shrink` has an effect on `p0` for `cy0`/`cy1`, and I don't know why. If `p0` is an _anchor_, then it should not get shrunk.

Reply:
I looked at this again, and `anchor` is the anchor point of _the colorbar_, while I was thinking it was the anchor on the parent Axes, but that's defined by `panchor`. So this is why I was confused by this calculation.

However, that means that `p0` doesn't seem to be calculated right, as it is calculating an anchor position in the parent Axes space (using `y0`/`y1`), but using `anchor_y`, which was used for the colorbar anchor position.

Reply:
In the test function, I will take `location=right` as an example

![demo](https://gitee.com/shawn96/publicFiles/raw/master/IMG.JPG)

the x0 y0 x1 y1 for the plot axes is [0.125, 0.11, 0.745, 0.88] (rounded to 2 digits) , blue region
the cx0 cy0 cx1 cy1 for cbar is [0.78, 0.22, 0.80, 0.61] , orange region

the p0 is the absolute y postion (red dot) of colorbar anchor against the plot, while anchor-y is a relative colorbar postion against the parent axes, 
thus `p0-y0 = (y1-y0) * anchor_y`.

Reply:
I assumed the y value for colorbar anchor  is the relative postion against the plot, not the absolute y position. Correct me if I am wrong.

Reply:
@ShawnChen1996  I think what would be convincing would be an example that shows clearly this version does close to the same as the version with `use_gridspec=False`, just to make sure the semantics are the same for `anchor` and `panchor`.    For completeness it would be good if the were tested at more than one anchor along all four sides.  

I know below we suggested not making an image test, but it would be helpful to see what the image looks like, even if it doesn't make it into the test.  I guess in situations like this, I actually lean towards adding an image test since its hard to tell by looking at the numbers if things are working.  But at least for review, it would be super helpful to see the image.

If the kwargs are being used the same as the old meaning of the kwargs, then we still have a documentation issue, but that shouldn't block this PR.